benchmarks/gups/gups.cpp

```diff
@@ -0,0 +1,201 @@
+  // (or the first entries of the permutation array if there
+  // are fewer indices than data elements)
+  Kokkos::View<Index*, DefaultHostMemorySpace> indices("indices", indicesCount);
+  for (Index i = 0; i < Index(indices.extent(0)); ++i) {
+    indices(i) = perm[i % perm.size()];
+  }
+
+  // Copy to the default space and return
+  IndexView dev_indices("dev_indices", indicesCount);
+  Kokkos::deep_copy(dev_indices, indices);
+  return dev_indices;
+}
+
+void run_gups(IndexView& indices, DataView& data, const Datum datum,
+              const bool performAtomics) {
+  if (performAtomics) {
+    Kokkos::parallel_for(
+        "bench-gups-atomic", indices.extent(0), KOKKOS_LAMBDA(const Index i) {
+          Kokkos::atomic_fetch_xor(&data[indices[i]], datum);
+        });
+  } else {
+    Kokkos::parallel_for(
+        "bench-gups-non-atomic", indices.extent(0),
+        KOKKOS_LAMBDA(const Index i) { data[indices[i]] ^= datum; });
+  }
+
+  Kokkos::fence();
+}
+
+enum class AccessPattern { random, permutation };
+
+int run_benchmark(const Index indicesCount, const Index dataCount,
+                  const int repeats, const bool useAtomics,
+                  const AccessPattern pattern) {
+  constexpr auto arbitrary_seed = 20230913;
+  RNG rng(arbitrary_seed);
+
+  printf("Reports fastest timing per kernel\n");
+  printf("Creating Views...\n");
+
+  printf("Memory Sizes:\n");
+  printf("- Elements:      %15" PRIu64 " (%12.4f MB)\n",
+         static_cast<uint64_t>(dataCount),
+         1.0e-6 * ((double)dataCount * (double)sizeof(Datum)));
+  printf("- Indices:       %15" PRIu64 " (%12.4f MB)\n",
+         static_cast<uint64_t>(indicesCount),
+         1.0e-6 * ((double)indicesCount * (double)sizeof(Index)));
+  printf(" - Atomics:      %15s\n", (useAtomics ? "Yes" : "No"));
+  printf("Benchmark kernels will be performed for %d iterations.\n", repeats);
+
+  printf(HLINE);
+
+  DataView dev_data("data", dataCount);
+  Datum datum = -1;
+
+  double gupsTime = 0.0;
+
+  printf("Initializing Data...\n");
+  Kokkos::View<Datum*, DefaultHostMemorySpace> data("data", dataCount);
+  Kokkos::parallel_for(
+      "init-data",
+      Kokkos::RangePolicy<Kokkos::DefaultExecutionSpace>(0, dataCount),
+      KOKKOS_LAMBDA(const int i) { dev_data[i] = 10101010101; });
+
+  printf("Starting benchmarking...\n");
+
+  for (Index k = 0; k < repeats; ++k) {
+    IndexView dev_indices;
+    switch (pattern) {
+      case AccessPattern::random: {
+        dev_indices = randomized_indices(indicesCount, dataCount, rng);
+        break;
+      }
+      case AccessPattern::permutation: {
+        dev_indices = permuted_indices(indicesCount, dataCount, rng);
+        break;
+      }
+      default: {
+        throw std::runtime_error("unexpected mode");
+      }
+    }
+
+    auto start = Clock::now();
+    run_gups(dev_indices, dev_data, datum, useAtomics);
+    gupsTime += Duration(Clock::now() - start).count();
+  }
+
+  Kokkos::deep_copy(data, dev_data);
```

Comment:
There is no point on copying the data back to the host.  We don't do anything with it and it goes out of scope when the function returns.